sql/pgwire: TestConn failed

[(sql/pgwire).TestConn failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1874520&tab=buildLog) on [release-20.1@ba27359b2de3dd1f3f57f171be4857966613f99b](https://github.com/cockroachdb/cockroach/commits/ba27359b2de3dd1f3f57f171be4857966613f99b):

```
I200414 05:28:55.542933 958 sqlmigrations/migrations.go:714  [n1] schema change job migration completed
I200414 05:28:55.543622 55 sql/pgwire/conn_test.go:219  executed query: SELECT t.oid, CASE WHEN nsp.nspname IN ('pg_catalog', 'public') THEN t.typname ELSE (nsp.nspname || '.') || t.typname END FROM pg_type AS t LEFT JOIN pg_type AS base_type ON t.typelem = base_type.oid LEFT JOIN pg_class AS base_cls ON base_type.typrelid = base_cls.oid LEFT JOIN pg_namespace AS nsp ON t.typnamespace = nsp.oid WHERE ((t.typtype IN ('b', 'p', 'r', 'e')) AND ((base_type.oid IS NULL) OR (base_type.typtype IN ('b', 'p', 'r'))))
I200414 05:28:55.543654 55 sql/pgwire/conn_test.go:200  advancing Sync
I200414 05:28:55.543660 979 sql/pgwire/conn_test.go:643  pgx log [info] Query - map[args:[] rowCount:%!s(int=70) sql:select t.oid,
	case when nsp.nspname in ('pg_catalog', 'public') then t.typname
		else nsp.nspname||'.'||t.typname
	end
from pg_type t
left join pg_type base_type on t.typelem=base_type.oid
left join pg_class base_cls ON base_type.typrelid = base_cls.oid
left join pg_namespace nsp on t.typnamespace=nsp.oid
where (
	  t.typtype in('b', 'p', 'r', 'e')
	  and (base_type.oid is null or base_type.typtype in('b', 'p', 'r'))
	) time:38.565293ms]
I200414 05:28:55.544256 959 sql/event_log.go:132  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:127.236.229.48:38515 Attrs: Locality: ServerVersion:19.2-18 BuildTag:v20.1.0-beta.4-423-gba27359 StartedAt:1586842135312347391 LocalityAddress:[] ClusterName: SQLAddress:127.236.229.48:32875} ClusterID:f94b9383-091b-4697-8726-3281480ed03a StartedAt:1586842135312347391 LastUp:1586842135312347391}
I200414 05:28:55.563318 55 sql/pgwire/conn_test.go:219  executed query: SELECT t.oid, t.typname FROM pg_type AS t JOIN pg_type AS base_type ON t.typelem = base_type.oid WHERE (t.typtype = 'b') AND (base_type.typtype = 'e')
I200414 05:28:55.563350 55 sql/pgwire/conn_test.go:200  advancing Sync
I200414 05:28:55.563563 979 sql/pgwire/conn_test.go:643  pgx log [info] Query - map[args:[] rowCount:%!s(int=1) sql:select t.oid, t.typname
from pg_type t
  join pg_type base_type on t.typelem=base_type.oid
where t.typtype = 'b'
  and base_type.typtype = 'e' time:19.361839ms]
I200414 05:28:55.567991 55 sql/pgwire/conn_test.go:219  executed query: SELECT t.oid, t.typname, t.typbasetype FROM pg_type AS t JOIN pg_type AS base_type ON t.typbasetype = base_type.oid WHERE (t.typtype = 'd') AND (base_type.typtype = 'b')
I200414 05:28:55.569128 55 sql/pgwire/conn_test.go:200  advancing Sync
I200414 05:28:55.569391 979 sql/pgwire/conn_test.go:643  pgx log [info] Query - map[args:[] rowCount:%!s(int=1) sql:select t.oid, t.typname, t.typbasetype
from pg_type t
  join pg_type base_type on t.typbasetype=base_type.oid
where t.typtype = 'd'
  and base_type.typtype = 'b' time:5.095352ms]
I200414 05:28:55.574434 55 sql/pgwire/conn_test.go:219  executed query: SELECT t.oid, t.typname FROM pg_type AS t JOIN pg_class AS cls ON t.typrelid = cls.oid WHERE (t.typtype = 'c') AND (cls.relkind = 'c')
I200414 05:28:55.574451 55 sql/pgwire/conn_test.go:200  advancing Sync
I200414 05:28:55.574546 979 sql/pgwire/conn_test.go:643  pgx log [info] Query - map[args:[] rowCount:%!s(int=1) sql:select t.oid, t.typname
from pg_type t
	join pg_class cls on t.typrelid=cls.oid
where t.typtype = 'c'
	and cls.relkind='c' time:4.69727ms]
I200414 05:28:55.575601 55 sql/pgwire/conn_test.go:212  stop wait at query: SELECT 1
I200414 05:28:55.575898 979 sql/pgwire/conn_test.go:643  pgx log [info] Exec - map[args:[] commandTag: sql:select 1 time:1.302001ms]
I200414 05:28:55.576239 979 sql/pgwire/conn_test.go:643  pgx log [info] Exec - map[args:[] commandTag: sql:select 2 time:307.084µs]
I200414 05:28:55.576733 979 sql/pgwire/conn_test.go:643  pgx log [info] Exec - map[args:[] commandTag: sql:p1 time:89.548µs]
I200414 05:28:55.577147 979 sql/pgwire/conn_test.go:643  pgx log [info] Exec - map[args:[] commandTag: sql:select 4; select 5; select 6; time:378.276µs]
I200414 05:28:55.578065 979 sql/pgwire/conn_test.go:643  pgx log [info] Exec - map[args:[] commandTag: sql:select 9 time:54.74µs]
I200414 05:28:55.578764 979 sql/pgwire/conn_test.go:643  pgx log [info] Exec - map[args:[] commandTag: sql:bogus statement failing to parse time:655.85µs]
I200414 05:28:57.588197 30 gossip/gossip.go:1527  [n1] node has connected to cluster via gossip
I200414 05:28:57.588423 30 kv/kvserver/stores.go:266  [n1] wrote 0 node addresses to persistent storage
I200414 05:29:00.579058 979 sql/pgwire/conn_test.go:643  pgx log [info] closed connection - map[]
I200414 05:29:00.579175 55 util/stop/stopper.go:539  quiescing
--- FAIL: TestConn (5.39s)
    conn_test.go:183: read tcp 127.0.0.1:40616->127.0.0.1:39173: i/o timeout

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestConn PKG=./pkg/sql/pgwire TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

Related:
- #46252 sql/pgwire: TestConn failed  [A-sql-pgwire](https://api.github.com/repos/cockroachdb/cockroach/labels/A-sql-pgwire) [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-master](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-master)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestConn.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/pgwire/conn.go
+++ b/pkg/sql/pgwire/conn.go
@@ -212,6 +212,8 @@ func (c *conn) serveImpl(
 	authOpt authOptions,
 	stopper *stop.Stopper,
 ) {
+	defer func() { _ = c.conn.Close() }()
+
 	ctx = logtags.AddTag(ctx, "user", c.sessionArgs.User)
 
 	inTestWithoutSQL := sqlServer == nil
@@ -220,7 +222,6 @@ func (c *conn) serveImpl(
 		authLogger = sqlServer.GetExecutorConfig().AuthLogger
 		sessionStart := timeutil.Now()
 		defer func() {
-			_ = c.conn.Close()
 			if c.authLogEnabled() {
 				authLogger.Logf(ctx, "session terminated; duration: %s", timeutil.Now().Sub(sessionStart))
 			}

